pkg/controller/csr/csr_controller.go

```diff
@@ -0,0 +1,321 @@
+		}
+		reqLogger.Info("Approved CSR with name : %v.", csr.Name)
+
+		certificatePEM, err := certificateManager.SignCertificate(certificateTemplate, certificateTemplate.PublicKey)
+		if err != nil {
+			reqLogger.Error(err, "error signing certificate request")
+			return reconcile.Result{}, err
+		}
+		updatedCSR.Status.Certificate = certificatePEM
+
+		_, err = r.certificatesClient.CertificateSigningRequests().UpdateStatus(ctx, updatedCSR, metav1.UpdateOptions{})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Signed CSR with name : %v.", csr.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+type systemError struct {
+	message string
+}
+
+func (v systemError) Error() string {
+	return v.message
+}
+
+func (r ReconcileCSR) validate(ctx context.Context, csr *certificatesv1.CertificateSigningRequest) (*x509.Certificate, error) {
+	firstBlock, restBlocks := pem.Decode(csr.Spec.Request)
+	if firstBlock == nil {
+		return nil, fmt.Errorf("cannot parse certificate request for CSR with name %s", csr.Name)
+	}
+	if len(restBlocks) != 0 {
+		return nil, fmt.Errorf("unexpected (multiple) pem blocks for CSR with name %s", csr.Name)
+	}
+
+	certificateRequest, err := x509.ParseCertificateRequest(firstBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	if err = certificateRequest.CheckSignature(); err != nil {
+		return nil, err
+	}
+
+	// The CSRName is formatted as "<secretName>:<podName>".
+	secretName := strings.Split(csr.Name, ":")[0]
+
+	// Validate whether this is a CSR we monitor at all.
+	asset, ok := r.allowedTLSAssets[secretName]
+	if !ok {
+		return nil, fmt.Errorf("this controller is not configured to sign secretName: %s", secretName)
+	}
+
+	// Validate whether the requestor of the CSR is registered for the given CSR
+	if fmt.Sprintf("system:serviceaccount:%s:%s", asset.serviceaccountNamespace, asset.serviceaccountName) != csr.Spec.Username {
+		return nil, fmt.Errorf("invalid requestor %s for CSR with name %s", csr.Spec.Username, csr.Name)
+	}
+
+	// Validate whether the DNS names are permitted for the request.
+	for _, name := range append(certificateRequest.DNSNames, certificateRequest.Subject.CommonName) {
+		var found bool
+		for _, valid := range asset.validDNSNames {
+			if valid == name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("invalid dns name \"%s\" found in CSR with name %s", name, csr.Name)
+		}
+	}
+
+	if len(certificateRequest.IPAddresses) == 1 {
+		// Validate the IP address that is requested in the CSR matches the pod IP of the requestor.
+		label, ok := csr.Labels[render.AppLabelName]
+		if !ok {
+			return nil, fmt.Errorf("unable to find expected k8s-app label for CSR with name %s", csr.Name)
+		}
+		podList := &corev1.PodList{}
+		err = r.client.List(ctx, podList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				render.AppLabelName: label,
+			}),
+			Namespace: asset.serviceaccountNamespace})
```

Comment:
if we're going to spread this across multiple lines, we should do so for the ending brackets as well:

```go
  Namespace: asset.serviceaccountNamespace,
 })
```

Or just put it all on one line because it doesn't actually seem _that_ long.